### DISCUSSION TITLE
Remapping Mouse Actions

### DISCUSSION THREAD

**Discussion by 00-kat** [ghostty-org/ghostty#3848] [Feature Requests, Ideas]
It would be nice to be able to remap mouse actions such as ctrl+triple click to select the last command output and ctrl+alt drag for block selection.  I also remember some people wanting to use alt+\<action\> (which is what iTerm2 apparently uses) to bypass mouse grab, as opposed to the current shift+\<action\>.

Edit: it would also be nice to map mouse actions to the normal actions that can currently only be used with the keyboard (which would help with things like #3783), as well as being able to map keyboard shortcuts to mouse actions.

Also, I'm not sure if this is relevant for this issue, but is there any list of mouse actions anywhere?  It would likely be useful should this feature request be accepted.

**Comment by SPYFF** [comment]
You can only remap keys listed [here](https://github.com/ghostty-org/ghostty/blob/main/src/input/key.zig#L255). So far, there is no option to remap mouse actions, however, does not looks like a hard change. Hopefully there are no platform specific issues regarding this.

**Reply by 00-kat** [reply]
I am aware; this was a feature request.  Thanks anyway, though.

**Comment by Igneous** [comment]
gnome and a lot of linux WM/DEs will clobber alt as well (as a hotkey for moving windows), this would be killer if it were rebindable.

Default behavior on gnome-terminal for block select is just ctrl.

**Reply by Igneous** [reply]
FYI: https://github.com/ghostty-org/ghostty/blob/f95f636f1fff1cf449efa79f3f42dcb8d9bbfce5/src/surface_mouse.zig#L126

nuke the `and mods.alt` and probably the test blocks below.

**Comment by nonetrix** [comment]
I would like to make ctrl+mouse zoom the terminal in and out, that is another good use case

**Reply by SPYFF** [reply]
Yes this is my only blocker right now. This is the default bind in every terminal emulator with Gnome VTE widget.

**Reply by tsugliani** [reply]
I would love to use something similar (mouse scroll remapping for font size increase/decrease) too:

```
keybind = ctrl+mouse_down=decrease_font_size:1
keybind = ctrl+mouse_up=increase_font_size:1
```

**Reply by unphased** [reply]
i like unifying ux across apps: it's pretty important that the same interactions can work across terminal and other apps, ctrl+mousewheel is one of these. It's extremely intuitive and an easy win feature for ghostty. Ideally we can have it configurable instead of hardcoded but in this particular case hardcoded would still be a step forward.

**Comment by 00-kat** [comment]
Some other relevant conversation: https://github.com/ghostty-org/ghostty/issues/4404#issuecomment-2676933597.

**Comment by jcollie** [comment]
Moving this from #4404 

Here's a potential syntax for a general mouse button action config. The performable prefix is mainly meant for use with the open-url or copy-url actions which would make that action a no-op if there's no URL underneath the click (or similarly no selection). A subsequent action could then open a context menu.

```
mouse-button-action = (performable:)?(ctrl|cmd|alt|super)*(left|right|middle|button-\d+)\+(press|release)(\+(single|double|triple|[1-9][0-9]*))?=(open-menu|open-url|copy-url|copy|select-and-copy|paste|select-word|select-line|unbind)
```

The difference between copy and select-and-copy is that select-and-copy will select a word if the click is not on an existing selection.

Some examples:
```
mouse-button-action = performable:right+press+single=copy-url
mouse-button-action = right+press+single=select-and-copy
mouse-button-action = ctrl+button-4+press+single=increase_font_size:2
mouse-button-action = ctrl+button-5+press+single=decrease_font_size:2
```

**Reply by 00-kat** [reply]
It might be useful to have a way to force a keybinding to work even if the mouse is captured by the application, à la Kitty's [`grabbed` and `ungrabbed`][kitty].

[kitty]: https://sw.kovidgoyal.net/kitty/conf/#mouse-actions

**Reply by itpropro** [reply]
Is this already implemented or just an idea?

**Reply by pluiedev** [reply]
Just an idea.

**Comment by jasalt** [comment]
It would be useful to be able to change the ctrl link hover modifier to something else like ctrl+alt as a workaround to avoid flashing the link target tooltip over the prompt.

By default using standard Emacs style hotkeys like `ctrl+a` (beginning of line) `ctrl+e` (end of line) `ctrl+p` (previous history entry) `ctrl+n` (next history entry) with bash/zsh leads to bad experience in Ghostty every time mouse cursor ends up on top of a link, which is happening few times a week to me.

https://github.com/user-attachments/assets/71e67e13-cbc7-4bbf-a0ae-e81342eb7105

The tooltip obstructs the prompt every time when using these keybindings. Related (closed) hyperlink tooltip discussion https://github.com/ghostty-org/ghostty/discussions/2654.

**Comment by carebare47** [comment]
Hey all,

I've made a start on this

Implemented:
  - Ctrl+scroll to zoom (font size increase/decrease)
  - Binding mouse buttons/scroll to any existing keybind action
  - Modifier+click combos (e.g. ctrl+right for URL copy)
  - Middle-click remapping
  - Extra buttons (mouse button_4 - button_11, but I don't have a fancy mouse so haven't been able to test this properly)
  - Click count parsing (double/triple syntax)
  - unbind and clear to remove bindings
  - Config syntax mirroring keybind style

  Not implemented:
  - Multi-click bindings don't actually fire at runtime (click_count always 1, explanation [here](https://github.com/carebare47/ghostty/blob/e46af256593485875d564696c70eafb4faa89da3/src/Surface.zig#L3772-L3774))
  - No press vs release distinction (@jcollie's proposal)
  - No performable: prefix for conditional actions (e.g. open-url only if URL present)
  - No mouse-specific actions (select-word, select-line, select-and-copy, open-menu)
  - No drag/block-selection remapping (ctrl+alt drag modifier can't be changed)
  - No grabbed/ungrabbed mode (force bindings when app captures mouse, like Kitty)
  - No link hover modifier remapping (ctrl+hover tooltip conflict)
  - No keyboard-to-mouse-action mapping (reverse direction)

Hopefully I'm allowed to drop a small nudge here, at the very least if someone approved wants to work on this then maybe they'd find it helpful to pick up where I've left it. Vouch request [here](https://github.com/ghostty-org/ghostty/discussions/11874), contains a link to the current implementation (PR to my own fork for now) - can probably find some time to expand this if the current semi-limited functionality is the blocker for acceptance/merge. Didn't want to do a comprehensive job of features that I don't personally need until I've got some kind of assurance that it would be merged upstream if I do

**Comment by carebare47** [comment]
Sorry for the slow response all

@mitchellh noted in my vouch request that the syntax needs approval here first, so here's a more concrete proposal

What's implemented so far mirrors `keybind`: `mouse-bind = [mods+]trigger[:count]=action`

e.g:
`ctrl+scroll_up=increase_font_size:1`
`middle=paste_from_clipboard`
`ctrl+right=copy_url_to_clipboard`
`ctrl+left:double=toggle_fullscreen` (full form with mods, count and action - count parses but doesn't fire yet)

Triggers are `left`/`right`/`middle`/`button_4..11`/`scroll_(up|down|left|right)`
Any keybind action is valid, no default bindings so existing behavior is untouched.

Differences from @jcollie's proposal:
- press vs release: I only handle press. Is release binding needed for v1, or acceptable as a follow-up?
- `performable:` prefix (no-op if no URL/selection under cursor): not implemented. Compatible with either syntax, could be a follow-up or included in a v1
- click counts: I parse `left:double`/`left:triple`/`left:2` (@jcollie's +double suffix would work equally well, no strong preference) - will wire up the runtime side as part of the PR
- mouse-specific actions (select-word, select-line, open-menu): not implemented. The action namespace is shared with keybind so they could be added later without syntax changes

Happy to rework the syntax to whatever's approved, mostly looking for a yes/no on the general shape

**Reply by TSheahan** [reply]
I'd love to see this go through. I want to configure right-click paste for Grok Build's TUI the way I already can in Kitty, without turning mouse reporting off for left-click / wheel / hover.

On the shape carebare47 asked about: yes to `mouse-bind = [mods+]trigger[:count]=action`, no default binds. That matches `keybind` and leaves current behavior alone until someone opts in.

The one first-slice question I care about is press vs release. Today's `right-click-action = paste` is a click (release). Press is what starts extend-select. If v1 is press-only, I don't think you can say "paste on right click" without also eating the press that clobbers the selection. Kitty splits those (`press` no-op, `click` paste). I'd rather see `:press` / `:release` (or an implicit click) in v1 than try to retrofit it later. `performable:` and mouse-only actions (`select-word`, `open-menu`) can wait.

jcollie's `right+press+single` reads to me as the same trigger with an explicit phase, not a second language. If v1 ships `mouse-bind` as proposed, that form still looks like a clean v2 addition (`right:release=paste_from_clipboard`, optional `performable:` prefix) instead of a rewrite. I would not block v1 on unifying the two writeups.

Grabbed vs ungrabbed (host bind while the app has DECSET mouse) feels like a later qualifier on the same trigger, not a v1 feature. Happy to talk through that once the grammar exists.

No strong preference on `:double` vs `+double` for count.

**Comment by JayToltTech** [comment]
Also need this.  Hitting it daily in opencode: with mouse capture on the wheel scrolls the conversation, which is what I want, but it also swallows `cmd+left` on links, so I'm stuck reaching for `shift+cmd+left` for everything.  The kitty-style 'grabbed' qualifier is the piece I'm after -- wheel and click stay captured, links don't have to.

Appreciate the work so far @carebare47.  Subscribed and looking forward to seeing this land.

**Comment by carebare47** [comment]
Thanks @TSheahan and @JayToltTech for the input

Taking that on board plus @jcollie's original proposal, here's a revised version that tries to merge the two:

`mouse-bind = [prefix:][mods+]button[+press|+release][+double|+triple|+N]=action`

- kept the name `mouse-bind` since it covers scroll as well as buttons, and matches `keybind`
- switched click count from `:double` to `+double`
- press/release is optional and defaults to press. You can bind both on the same button. If a click matches any binding, the whole click is handled by ghostty (neither the press nor the release gets passed on), so you don't get half a click
- press/release and click counts don't apply to scroll
- `performable:` works the same as it does for `keybind` - if the action can't be performed (e.g. no URL under the cursor) the click falls through to the normal behaviour
- a matching binding takes priority over `right-click-action`/`middle-click-action`

e.g.
```
mouse-bind = ctrl+scroll_up=increase_font_size:1
mouse-bind = middle=paste_from_clipboard
mouse-bind = right+release=paste_from_clipboard
mouse-bind = performable:ctrl+right=copy_url_to_clipboard
mouse-bind = ctrl+left+double=select_all
```

One thing I think needs deciding for v1 is what happens when a program running in the terminal has mouse reporting turned on (vim with `set mouse=a`, tmux with `mouse on`, opencode etc). At the moment my branch checks bindings before mouse reporting, so a bound ctrl+scroll always gets stolen from vim.

I'd propose bindings don't fire while mouse reporting is on by default, plus a kitty-style prefix to opt in (e.g. `grabbed:ctrl+scroll_up=increase_font_size:1`, name open to suggestions). I think the prefix maybe should be in v1 rather than a follow-up, because scroll doesn't have a shift bypass like clicks do - without it there'd be no way to zoom while vim has the mouse (but again open to input here)

(fwiw `right-click-action` fires on press today rather than release ([here](https://github.com/ghostty-org/ghostty/blob/4ae9f1a2de5484de3d6a13fe03676b8853b9c41c/src/Surface.zig#L4100)), but I'm happy to include release in v1 anyway since it's a small addition)

@jcollie does this look reasonable to you? Happy to go with whatever you and mitchellh prefer. I'll wire up release, multi-click and `performable:` as part of the PR

(AI disclosure: I used Claude to help check this against the current code and tidy up the wording)
